### PR TITLE
GH-768: Implement copy function from on-prem to Cloud

### PR DESCRIPTION
**Pull request by rexwangcc**
### Purpose

- https://broadinstitute.atlassian.net/browse/GH-768

### Changes

- [x] Adapt and copy some existing functions from WFL for the GCS interactions.
- [x] Set up Linter/Formatter for the code base.
- [x] Refactor some code organization.
- [x] Implement message peek func in case we need it later.
- [x] Unit tests.
- [x] Integration tests.

### Review Instructions

- `clojure -A:test` for unit testing!
- Connect to VPN and run `clojure -A:integration` for integration testing!
(You could monitor the [bucket](https://console.cloud.google.com/storage/browser/broad-gotc-dev-zero-test?forceOnBucketsSortingFiltering=false&organizationId=548622027621&project=broad-gotc-dev&pli=1) for the uploaded object during the integration test if you are interested)

### CODE REVIEW

**Comment by rexwangcc** [comment]
>maybe we can start a library and reference as a sub module?

Highly up for it otherwise we will run into inconsistency problems as the code grow. Speaking of that, we should also talk about choosing mono-repo or multi-repo organization for WFL asap.

File: src/util/gcs.clj [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+         :body
+         (json/read-str :key-fn keyword))))
+  ([file bucket object]
+   (upload-file file bucket object (once/get-auth-header!)))
+  ([file url]
+   (apply upload-file file (parse-gs-url url))))
+
+(comment
+  (list-objects "broad-gotc-dev-zero-test")
+  (upload-file "project.clj" "broad-gotc-dev-zero-test" "junk-test.clj"))
```
Comment (samanehsan) [inline src/util/gcs.clj:81]: I think these can be turned into unit tests!

File: deps.edn [hunk 2]
```diff
@@ -1,18 +1,25 @@
...
-           :uberdeps {:extra-deps {uberdeps {:mvn/version "0.1.10"}}
-                      :main-opts  ["-m" "uberdeps.uberjar"]}}
+ :aliases {:test      {:extra-paths ["test"]
+                       :extra-deps  {lambdaisland/kaocha {:mvn/version "1.0.629"}
+                                     com.google.cloud/google-cloud-nio {:mvn/version "0.5.0"}}
+                       :main-opts   ["-m" "kaocha.runner"]}
+           :uberdeps  {:extra-deps {uberdeps {:mvn/version "0.1.10"}}
+                       :main-opts  ["-m" "uberdeps.uberjar"]}
+           :lint      {:main-opts ["-m" "cljfmt.main" "check"]}
+           :format    {:main-opts ["-m" "cljfmt.main" "fix"]}}
```
Comment (samanehsan) [inline deps.edn:24]: I'm not familiar with the differences between cljfmt and eastwood but it'll probably be easier to use the same one for both PTC & WFL!

File: src/util/misc.clj [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    (let [from (str ptc/the-name "@broadinstitute.org")
+          subject "This thing is from PTC service"]
+      (doto (-> (new SimpleEmail)
+                (.setFrom from)
+                (.setSubject subject)
+                (.setMsg message))
+        (add-to-list to-list)
+        (.setAuthentication from "fake-password")
+        (.setHostName "smtp.gmail.com")
+        (.send)))))
```
Comment (samanehsan) [inline src/util/misc.clj:47]: We could look into sending these emails asynchronously as a future optimization

File: deps.edn [hunk 4]
```diff
@@ -1,18 +1,25 @@
...
-           javax.servlet/servlet-api         {:mvn/version "2.5"}
-           org.apache.activemq/activemq-core {:mvn/version "5.7.0"}
-           org.slf4j/slf4j-simple            {:mvn/version "1.7.30"}
-           amperity/vault-clj                {:mvn/version "0.5.1"}
-           org.apache.commons/commons-email  {:mvn/version "1.5"}
-           com.taoensso/timbre               {:mvn/version "4.10.0"}}
+          {org.clojure/clojure                   {:mvn/version "1.10.1"}
+           org.clojure/data.json                 {:mvn/version "1.0.0"}
+           clj-time                              {:mvn/version "0.14.2"}
+           cljfmt                                {:mvn/version "0.6.7"}
```
Comment (ehigham) [inline deps.edn:7]: Is `cljfmt` a runtime dependence or just required for development? Can it be made a dependence of the `lint` and `format` aliases?

File: src/util/once.clj [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+(defonce the-cached-credentials (delay (new-user-credentials)))
+
+(defn get-auth-header!
+  "Return a valid auth header, refresh the access token
+  under the hood if necessary."
+  []
+  (misc/bearer-token-header-for @the-cached-credentials))
+
+(comment
+  (get-auth-header!))
```
Comment (ehigham) [inline src/util/once.clj:37]: might be an idea to configure your editor to add a blank line at the end?

File: deps.edn (hunk 4 above)
Comment (rexwangcc) [reply deps.edn:7]: Good catch! Lemme see if we could move it to the aliases!!

File: src/util/once.clj (hunk 5 above)
Comment (rexwangcc) [reply src/util/once.clj:37]: Yeah, let me see how to do it...

File: deps.edn (hunk 2 above)
Comment (rexwangcc) [reply deps.edn:24]: Thanks, that makes sense! My understanding is that `eastwood` is just a linter (think about flake8), while `cljfmt` can both lint and format (think about yapf or black), which is why I went down with `cljfmt`. What do you think?

File: src/util/gcs.clj (hunk 1 above)
Comment (rexwangcc) [reply src/util/gcs.clj:81]: I haven't found a good way to mock the GCS without using https://github.com/googleapis/google-cloud-java/tree/master/google-cloud-clients/google-cloud-contrib/google-cloud-nio with GCloud Java Clients, so i only adopted some of the existing unit tests from the `WFL` repo that don't require mocking. I think for this specific one, an integration test that works with an actual bucket might be easier for us, WDYT?

File: src/util/misc.clj (hunk 3 above)
Comment (rexwangcc) [reply src/util/misc.clj:47]: 👍

File: src/start.clj [hunk 6]
```diff
@@ -72,18 +72,25 @@
...
+  ([connection queue task!]
+   (let [msg (edn/read-string (slurp "./test/data/test_msg.edn"))
+         title (get-in msg [:headers :message-id])
+         text  (walk/stringify-keys (:properties msg))]
+     (loop [counter 0
+            msg-state (atom nil)]
+       (produce connection queue title text)
+       (try (when-let [message (consume connection queue)]
+              (let [message (parse-message message)]
+                (reset! msg-state message)
```
Comment (rexwangcc) [inline src/start.clj:86]: `clojure.lang.PersistentArrayMap cannot be cast to clojure.lang.IAtom` :(

File: src/util/gcs.clj (hunk 1 above)
Comment (rexwangcc) [reply src/util/gcs.clj]: done!

File: deps.edn [hunk 7]
```diff
@@ -1,18 +1,27 @@
...
-                      :extra-deps  {lambdaisland/kaocha {:mvn/version "1.0.629"}}
-                      :main-opts   ["-m" "kaocha.runner"]}
-           :uberdeps {:extra-deps {uberdeps {:mvn/version "0.1.10"}}
-                      :main-opts  ["-m" "uberdeps.uberjar"]}}
+ :aliases {:test        {:extra-paths ["test"]
+                         :extra-deps  {lambdaisland/kaocha               {:mvn/version "1.0.629"}}
+                         :main-opts   ["-m" "kaocha.runner" "unit"]}
+           :integration {:extra-paths ["test"]
+                         :extra-deps  {lambdaisland/kaocha               {:mvn/version "1.0.629"}}
+                         :main-opts   ["-m" "kaocha.runner" "integration"]}
```
Comment (ehigham) [inline deps.edn:20]: Maybe we can just combine these into one and just pass the "kind" via the command line? (See https://github.com/broadinstitute/wfl/pull/43/files#diff-29bd8491c570ce25397243cf594a6245L77 for an example?)

Comment (rexwangcc) [reply deps.edn]: (Just stolen from your PR, thanks for the example 😄 )

**Pull request merged**